### PR TITLE
OSD

### PR DESCRIPTION
**Pull request by paoloo**
## What and Why
Add OSD schema and implement GetOSD and GetOSDs

## Verification Steps
- `mix test test/media/ver10/get_osds.exs`to run the unit tests
- `Onvif.Media.Ver10.GetOSDs.request` to List and get the token
- `Onvif.Media.Ver10.GetOSD.request` to Get

## Tested on
- HIKVISION: :heavy_check_mark:
- Dahua: :heavy_check_mark:
- AXIS: :heavy_check_mark: 
- UNIVIEW: :heavy_check_mark: (using the `Video Source Configuration` as parameter)

### CODE REVIEW

File: lib/media/ver10/get_osds.ex [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+    |> Enum.reduce([], fn raw_service, acc ->
+      case OSD.to_struct(raw_service) do
+        {:ok, service} ->
+          [service | acc]
+
+        {:error, changeset} ->
+          Logger.error("Discarding invalid service: #{inspect(changeset)}")
+          acc
+      end
+    end)
```
Comment (waranlogesh) [inline lib/media/ver10/get_osds.ex:39]:
```suggestion
    response = 
      xml_response_body
      |> parse(namespace_conformant: true, quiet: true)
      |> xpath(
        ~x"//s:Envelope/s:Body/trt:GetOSDsResponse/trt:OSDs"el
        |> add_namespace("s", "http://www.w3.org/2003/05/soap-envelope")
        |> add_namespace("trt", "http://www.onvif.org/ver10/media/wsdl")
        |> add_namespace("tt", "http://www.onvif.org/ver10/schema")
      )
      |> Enum.map(&OSD.parse/1)
      |> Enum.reduce([], fn raw_service, acc ->
        case OSD.to_struct(raw_osd) do
          {:ok, osd} ->
            [osd| acc]

          {:error, changeset} ->
            Logger.error("Discarding invalid osd: #{inspect(changeset)}")
            acc
        end
        {:ok, response}
      end)
```

Changed variable service to osd and made the return to be {:ok, response} to be consistent with other media APIs. You need to change the tests to handle new return type.

File: test/media/ver10/get_osds.exs [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+    test "should have no OSD available" do
+      xml_response = File.read!("test/media/ver10/fixtures/get_osds_empty.xml")
+
+      device = Onvif.Factory.device()
+
+      Mimic.expect(Tesla, :request, fn _client, _opts ->
+        {:ok, %{status: 200, body: xml_response}}
+      end)
+
+      osds = Onvif.Media.Ver10.GetOSDs.request(device)
```
Comment (waranlogesh) [inline test/media/ver10/get_osds.exs:59]: Thanks for the clear tests. I was able to find about the return type easily from here.

**Review by waranlogesh — APPROVED** [review]
Only fails in Uniview if we don't pass the video source token. We can debug this later. It passes in Axis, Hikvision and Dahua.
```
<?xml version="1.0" encoding="UTF-8"?>
<SOAP-ENV:Envelope xmlns:SOAP-ENV="http://www.w3.org/2003/05/soap-envelope" xmlns:SOAP-ENC="http://www.w3.org/2003/05/soap-encoding" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xmlns:xsd="http://www.w3.org/2001/XMLSchema" xmlns:tns="http://tempuri.org/tns.xsd" xmlns:tns1="http://www.onvif.org/ver10/topics" xmlns:c14n="http://www.w3.org/2001/10/xml-exc-c14n#" xmlns:wsu="http://docs.oasis-open.org/wss/2004/01/oasis-200401-wss-wssecurity-utility-1.0.xsd" xmlns:ds="http://www.w3.org/2000/09/xmldsig#" xmlns:wsse="http://docs.oasis-open.org/wss/2004/01/oasis-200401-wss-wssecurity-secext-1.0.xsd" xmlns:wsa="http://schemas.xmlsoap.org/ws/2004/08/addressing" xmlns:wsa5="http://www.w3.org/2005/08/addressing" xmlns:xmime5="http://tempuri.org/xmime5.xsd" xmlns:xop="http://www.w3.org/2004/08/xop/include" xmlns:tplt="http://www.onvif.org/ver10/plus/schema" xmlns:tt="http://www.onvif.org/ver10/schema" xmlns:wstop="http://docs.oasis-open.org/wsn/t-1" xmlns:wsrfr="http://docs.oasis-open.org/wsrf/r-2" xmlns:wsrfbf="http://docs.oasis-open.org/wsrf/bf-2" xmlns:tanae="http://www.onvif.org/ver20/analytics/wsdl/AnalyticsEngineBinding" xmlns:tanre="http://www.onvif.org/ver20/analytics/wsdl/RuleEngineBinding" xmlns:tan="http://www.onvif.org/ver20/analytics/wsdl" xmlns:tds="http://www.onvif.org/ver10/device/wsdl" xmlns:tes-cppb="http://www.onvif.org/ver10/events/wsdl/CreatePullPointBinding" xmlns:tes-e="http://www.onvif.org/ver10/events/wsdl/EventBinding" xmlns:tes-nc="http://www.onvif.org/ver10/events/wsdl/NotificationConsumerBinding" xmlns:tes-np="http://www.onvif.org/ver10/events/wsdl/NotificationProducerBinding" xmlns:tes-ppb="http://www.onvif.org/ver10/events/wsdl/PullPointBinding" xmlns:tes-pps="http://www.onvif.org/ver10/events/wsdl/PullPointSubscriptionBinding" xmlns:tev="http://www.onvif.org/ver10/events/wsdl" xmlns:tes-psmb="http://www.onvif.org/ver10/events/wsdl/PausableSubscriptionManagerBinding" xmlns:wsnt="http://docs.oasis-open.org/wsn/b-2" xmlns:tes-sm="http://www.onvif.org/ver10/events/wsdl/SubscriptionManagerBinding" xmlns:timg="http://www.onvif.org/ver20/imaging/wsdl" xmlns:tmd="http://www.onvif.org/ver10/deviceIO/wsdl" xmlns:tpl="http://www.onvif.org/ver10/plus/wsdl" xmlns:tptz="http://www.onvif.org/ver20/ptz/wsdl" xmlns:tr2="http://www.onvif.org/ver20/media/wsdl" xmlns:trc="http://www.onvif.org/ver10/recording/wsdl" xmlns:trp="http://www.onvif.org/ver10/replay/wsdl" xmlns:trt="http://www.onvif.org/ver10/media/wsdl" xmlns:trv="http://www.onvif.org/ver10/receiver/wsdl" xmlns:tse="http://www.onvif.org/ver10/search/wsdl" xmlns:ter="http://www.onvif.org/ver10/error">
  <SOAP-ENV:Header/>
  <SOAP-ENV:Body>
    <SOAP-ENV:Fault>
      <SOAP-ENV:Code>
        <SOAP-ENV:Value>SOAP-ENV:Sender</SOAP-ENV:Value>
        <SOAP-ENV:Subcode>
          <SOAP-ENV:Value>ter:InvalidArgVal</SOAP-ENV:Value>
        </SOAP-ENV:Subcode>
      </SOAP-ENV:Code>
      <SOAP-ENV:Reason>
        <SOAP-ENV:Text xml:lang="en">Argument Value Invalid</SOAP-ENV:Text>
      </SOAP-ENV:Reason>
      <SOAP-ENV:Detail/>
    </SOAP-ENV:Fault>
  </SOAP-ENV:Body>
</SOAP-ENV:Envelope>
```

File: lib/media/ver10/osd.ex [hunk 3]
```diff
@@ -0,0 +1,225 @@
...
+      field(:date_format, Ecto.Enum,
+        values: [
+          "M/d/yyyy": "M/d/yyyy",
+          "MM/dd/yyyy": "MM/dd/yyyy",
+          "dd/MM/yyyy": "dd/MM/yyyy",
+          "yyyy/MM/dd": "yyyy/MM/dd",
+          "yyyy-MM-dd": "yyyy-MM-dd",
+          "dddd, MMMM dd, yyyy ": "dddd, MMMM dd, yyyy ",
+          "MMMM dd, yyyy ": "MMMM dd, yyyy ",
+          "dd MMMM, yyyy": "dd MMMM, yyyy"
```
Comment (hammeraj) [inline lib/media/ver10/osd.ex:52]: this looks a little odd to me, I don't think we want to have string to atom conversion here. Instead maybe you can do a validator https://hexdocs.pm/ecto/Ecto.Changeset.html#validate_inclusion/4 with the expected values in a list.

Comment (hammeraj) [reply lib/media/ver10/osd.ex:52]: Same for time format

**Review by hammeraj — APPROVED** [review]

**Pull request merged**